fastscanner/adapters/candle/parquet.py

```diff
@@ -0,0 +1,165 @@
+import os
+import json
+import logging
+from calendar import monthrange
+from datetime import date, datetime, time, timedelta
+from typing import Dict, List
+from zoneinfo import ZoneInfo
+
+import httpx
+import pandas as pd
+import pyarrow as pa
+import pyarrow.dataset as ds
+import pyarrow.parquet as pq
+
+from fastscanner.pkg.localize import LOCAL_TIMEZONE_STR
+from . import config
+from .polygon import CandleCol, PolygonBarsProvider, split_freq
+
+logger = logging.getLogger(__name__)
+
+class ParquetBarsProvider(PolygonBarsProvider):
+    CACHE_DIR = os.path.join("data", "parquet_dataset")
+    tz: str = LOCAL_TIMEZONE_STR
+
+    def get(self, symbol: str, start: date, end: date, freq: str) -> pd.DataFrame:
+        _, unit = split_freq(freq)
+        keys = self._partition_keys_in_range(start, end, unit)
+
+        dataset_path = os.path.join(self.CACHE_DIR, symbol)
+        if not os.path.exists(dataset_path):
+            return pd.DataFrame(columns=list(CandleCol.RESAMPLE_MAP.keys()),
+                                index=pd.DatetimeIndex([], name=CandleCol.DATETIME)).tz_localize(self.tz)
+
+        filter_expr = (
+            ds.field("unit") == unit
+        ) & (
+            ds.field("partition_key").isin(keys)
+        )
+
+
+        try:
+            dataset = ds.dataset(dataset_path, format="parquet", partitioning="hive")
+            table = dataset.to_table(filter=filter_expr)
+            df = table.to_pandas()
+        except Exception as e:
+            logger.error(f"Failed to read from Parquet dataset: {e}")
+            return pd.DataFrame(columns=list(CandleCol.RESAMPLE_MAP.keys()),
+                                index=pd.DatetimeIndex([], name=CandleCol.DATETIME)).tz_localize(self.tz)
+
+        if df.empty:
+            return df
+
+        df[CandleCol.DATETIME] = pd.to_datetime(df[CandleCol.DATETIME], utc=True)
+        df = df.set_index(CandleCol.DATETIME).tz_convert(self.tz)
+
+        start_dt = datetime.combine(start, time.min).replace(tzinfo=ZoneInfo(self.tz))
+        end_dt = datetime.combine(end, time.max).replace(tzinfo=ZoneInfo(self.tz))
+        df = df.loc[start_dt:end_dt]
+
+        if freq in ("1min", "1h", "1d"):
+            return df
+
+        return df.resample(freq).aggregate(CandleCol.RESAMPLE_MAP).dropna()
+
+    def _cache(self, symbol: str, key: str, unit: str) -> pd.DataFrame:
+        dataset_path = os.path.join(self.CACHE_DIR, symbol)
+
+        if os.path.exists(dataset_path) and not self._is_expired(symbol, key, unit):
```

Comment:
We can also change the logic a bit. Instead of using key expiration, we can store the current [start,end] in the current_ranges.json file (check the comment below in the mark_expiration function). Then instead of is_expired we can call
```
missing_ranges = self._missing_ranges(symbol, start, end, freq)
```
Check below for the comments on the missing_ranges